src/back/spirv/layout.rs

```diff
@@ -0,0 +1,216 @@
+use spirv::*;
+
+pub(crate) struct Module {
+    pub(crate) physical_layout: PhysicalLayout,
+    pub(crate) logical_layout: LogicalLayout,
+}
+
+impl Module {
+    pub(crate) fn new(module: &crate::Module) -> Self {
+        let version: Word = (0x0u32 << 24)
+            | ((module.header.version.0 as u32) << 16)
+            | ((module.header.version.1 as u32) << 8)
+            | module.header.version.2 as u32;
+
+        Module {
+            physical_layout: PhysicalLayout::new(version, module.header.generator, 0x0u32),
+            logical_layout: LogicalLayout::new(),
+        }
+    }
+}
+
+pub(crate) struct PhysicalLayout {
+    magic_number: Word,
+    version: Word,
+    generator: Word,
+    pub(crate) bound: Word,
+    instruction_schema: Word,
+}
+
+impl PhysicalLayout {
+    pub(crate) fn new(version: Word, generator: Word, instruction_schema: Word) -> Self {
+        PhysicalLayout {
+            magic_number: MAGIC_NUMBER,
+            version,
+            generator,
+            bound: 0,
+            instruction_schema,
+        }
+    }
+
+    pub(crate) fn in_words(&self) -> Vec<Word> {
+        let mut words: Vec<Word> = vec![];
+
+        words.push(self.magic_number);
+        words.push(self.version);
+        words.push(self.generator);
+
+        match self.bound {
+            0 => panic!("Bound cannot be 0"),
+            _ => words.push(self.bound),
+        }
+
+        words.push(self.instruction_schema);
+        words
+    }
+}
+
+pub(crate) struct LogicalLayout {
+    pub(crate) capabilities: Vec<Word>,
```

Comment:
Well that depends on how we want to deal with the parsing. My philosophy is doing it in steps and doing it like so:
1) Parse the IR module, start by going through everything in a AST-like order, so we would reduce the amount of iterations over the IR. Now it doesn't have to comply to the order of instructions that SPIR-V expects. That is done in the `write_logical_layout` in `parser.rs`
2) After everything is parsed, loop through all those Instructions, but now order them in the way SPIR-V expects. That is done here in the `in_words`.

Benefit of doing this is that we are free in what order we process things, instead of immediately adding it to the output stream, and we are able to separate the logic of parsing with the actual structure of a SPIR-V module.

I am not yet sure how we could improve this without having these vectors at the moment.

Reply:
Ok, the problem I see with this approach is that it's basically introducing another IR layer. And given that this is SPIRV, which is semantically close to WGSL and Naga IR, having *another* layer here is akin to admitting that Naga IR failed.

We should try to write things down on the fly with a single pass over the module (plus whatever the `GlobalUsage` processor does). If that is not possible for some thing, then we should start asking ourselves if IR needs to be changed. And only then we'd fall back to having multiple passes.

Reply:
Great point made, I think we can definitely do one pass directly translating the IR to the SPIR-V. I am not yet sure how, but we could discuss that.

Reply:
Gave this idea some time to grow, but I'm not sure if there is a chance to have it single passed. 

So we are back at the topic why it is designed like this: the Functions arena is parsed first, due to loop efficiency. The vision I had, was that when looping through Functions and encountering Types, Constants, Variables etc. it would be checked if it is already parsed, if not, parse it. So we are removing the need to loop through each individual arena.

There are variables like Pointers, and maybe some other instructions that are generated on runtime, but need to be inserted in the Type Declarations. If we do it single pass, it wouldn't be possible as we don't know where the Type Declarations are in the one stream of Words.

I'm thinking about if the IR can be improved to help this case, but I am not sure at the moment. Do you think it should be possible to do it in one pass?

Reply:
let's talk about this on the chat